test_conformance/basic/test_fpmath.cpp

```diff
@@ -0,0 +1,381 @@
+extension,
+"__kernel void test_fp(__global TYPE *srcA, __global TYPE *srcB, __global TYPE *dst)\n"
+"{\n"
+"    int tid = get_global_id( 0 );\n",
+"    dst[tid] = srcA[tid] OP srcB[tid];\n",
+"}\n"};
+// clang-format on
+
+template <typename T> double toDouble(T val)
+{
+    if (std::is_same<cl_half, T>::value)
+        return HTF(val);
+    else
+        return val;
+}
+
+bool isHalfNan(cl_half v)
+{
+    // Extract FP16 exponent and mantissa
+    uint16_t h_exp = (v >> (CL_HALF_MANT_DIG - 1)) & 0x1F;
+    uint16_t h_mant = v & 0x3FF;
+
+    // NaN test
+    return (h_exp == 0x1F && h_mant != 0);
+}
+
+cl_half half_plus(cl_half a, cl_half b)
+{
+    return HFF(std::plus<float>()(HTF(a), HTF(b)));
+}
+
+cl_half half_minus(cl_half a, cl_half b)
+{
+    return HFF(std::minus<float>()(HTF(a), HTF(b)));
+}
+
+cl_half half_mult(cl_half a, cl_half b)
+{
+    return HFF(std::multiplies<float>()(HTF(a), HTF(b)));
+}
+
+template <typename T> struct TestDef
+{
+    const char op;
+    std::function<T(T, T)> ref;
+    std::string type_str;
+    size_t vec_size;
+};
+
+template <typename T>
+int verify_fp(std::vector<T> (&input)[2], std::vector<T> &output,
+              const TestDef<T> &test)
+{
+    auto &inA = input[0];
+    auto &inB = input[1];
+    for (int i = 0; i < output.size(); i++)
+    {
+        bool nan_test = false;
+
+        T r = test.ref(inA[i], inB[i]);
+
+        if (std::is_same<T, cl_half>::value)
+            nan_test = !(isHalfNan(r) && isHalfNan(output[i]));
+
+        if (r != output[i] && nan_test)
+        {
+            log_error("FP math test for type: %s, vec size: %zu, failed at "
+                      "index %d, %a '%c' %a, expected %a, get %a\n",
+                      test.type_str.c_str(), test.vec_size, i, toDouble(inA[i]),
+                      test.op, toDouble(inB[i]), toDouble(r),
+                      toDouble(output[i]));
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+template <typename T> void generate_random_inputs(std::vector<T> (&input)[2])
+{
+    RandomSeed seed(gRandomSeed);
+
+    if (std::is_same<T, float>::value)
+    {
+        auto random_generator = [&seed]() {
+            return get_random_float(-MAKE_HEX_FLOAT(0x1.0p31f, 0x1, 31),
+                                    MAKE_HEX_FLOAT(0x1.0p31f, 0x1, 31), seed);
+        };
+        for (auto &v : input)
+            std::generate(v.begin(), v.end(), random_generator);
+    }
+    else if (std::is_same<T, double>::value)
+    {
+        auto random_generator = [&seed]() {
+            return get_random_double(-MAKE_HEX_DOUBLE(0x1.0p63, 0x1LL, 63),
+                                     MAKE_HEX_DOUBLE(0x1.0p63, 0x1LL, 63),
+                                     seed);
+        };
+        for (auto &v : input)
+            std::generate(v.begin(), v.end(), random_generator);
+    }
+    else
+    {
+        auto random_generator = [&seed]() {
+            return get_random_float(HFF(-MAKE_HEX_FLOAT(0x1.0p15f, 0x1, 15)),
```

Comment:
I don't think this produces a good set of test values.  From a quick inspection of the generated input values for half, the vast majority of them seem to have the sign bit set for example.  Perhaps we need a dedicated `get_random_half` helper function?

Reply:
Good find, corrected. 

> Perhaps we need a dedicated get_random_half helper function?

probably, this is only an impression but it seems like most of test cases uses customized input values

Reply:
It's a bit better now, but I still have some concerns on whether the generated test data for fp16 is meaningful enough:

1. Most randomly generated values are > 1024, so most of the test values don't have any interesting fractional part.
2. For multiplications, the randomly generated input operands are that large that the resulting product is almost always infinity.  Perhaps it makes sense to fine-tune the fp16 generation to generate smaller numbers more often?
3. Behavior near 0 isn't covered, but the float/double cases also seem to have this problem, so we could ignore that for now.

Reply:
Actually point 1 applies to float/double too indeed, so like point 3 we can leave that as a future improvement.  But point 2 seems specific to fp16: because of the very limited range of fp16, it's much easier to hit +/- infinity.  I just checked, for float the multiplication test doesn't produce infinities.

Reply:
> I just checked, for float the multiplication test doesn't produce infinities.

of course, that's probably because float range (`+/- 0x1.0p31f`) leaves enough space for multiplication meanwhile half range is at the edge of precision. I think we should decrease it to `+/- 0x1.0p8f`